src/material-experimental/mdc-list/list-base.ts

```diff
@@ -77,20 +86,113 @@ export abstract class MatListItemBase implements AfterContentInit, OnDestroy, Ri
     this._ngZone.runOutsideAngular(() => {
       this._subscriptions.add(this.lines.changes.pipe(startWith(this.lines))
           .subscribe((lines: QueryList<ElementRef<Element>>) => {
-            this._element.nativeElement.classList
+            this._element.classList
                 .toggle('mat-mdc-list-item-single-line', lines.length <= 1);
             lines.forEach((line: ElementRef<Element>, index: number) => {
               toggleClass(line.nativeElement,
                   'mdc-list-item__primary-text', index === 0 && lines.length > 1);
               toggleClass(line.nativeElement, 'mdc-list-item__secondary-text', index !== 0);
             });
-            setLines(lines, this._element, 'mat-mdc');
+            setLines(lines, this._elementRef, 'mat-mdc');
           }));
     });
   }
+}
+
+@Directive()
+/** @docs-private */
+export abstract class MatListBase {
+  @HostBinding('class.mdc-list--non-interactive')
+  _isNonInteractive: boolean = true;
+}
+
+@Directive()
+export abstract class MatInteractiveListBase extends MatListBase
+    implements AfterViewInit, OnDestroy {
+  @HostListener('keydown', ['$event'])
+  _handleKeydown(event: KeyboardEvent) {
+    const index = this._indexForElement(event.target as HTMLElement);
+    this._foundation.handleKeydown(
+        event, this._itemAtIndex(index)._element === event.target, index);
+  }
+
+  @HostListener('click', ['$event'])
+  _handleClick(event: MouseEvent) {
+    this._foundation.handleClick(this._indexForElement(event.target as HTMLElement), false);
+  }
+
+  @HostListener('focusin', ['$event'])
+  _handleFocusin(event: FocusEvent) {
+    this._foundation.handleFocusIn(event, this._indexForElement(event.target as HTMLElement));
+  }
+
+  @HostListener('focusout', ['$event'])
+  _handleFocusout(event: FocusEvent) {
+    this._foundation.handleFocusOut(event, this._indexForElement(event.target as HTMLElement));
+  }
+
+  @ContentChildren(MatListItemBase) _items: QueryList<MatListItemBase>;
+
+  protected _adapter: MDCListAdapter = {
+    getListItemCount: () => this._items.length,
+    listItemAtIndexHasClass:
+        (index, className) => this._itemAtIndex(index)._element.classList.contains(className),
```

Comment:
It looks like you're repeating `this._itemAtIndex(index)._element` a lot. Would it make sense to have `_elementAtIndex` that does the lookup and returns the element?

Reply:
Yeah probably, that reminds me I also meant to refactor it so we're not calling `toArray` constantly